projects/flow.groups-2D-random-walk/src/project.py

```diff
@@ -0,0 +1,152 @@
+import flow
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+# Helper functions
+def get_first_replica(jobs, first_index=1):
+    try:
+        job = next(filter(lambda job: job.sp.replica == first_index, jobs))
+    except StopIteration:
+        raise RuntimeError("First replica not found.")
+    return job
+
+
+def positions_generator(jobs):
+    for job in jobs:
+        with job.data:
+            yield job.data.positions
+
+
+def aggregate_cond_true(doc_attr):
+    """Create a condition which is true if all jobs have doc_attr as truthy."""
+
+    def true_for_all_docs(*jobs):
+        return all(job.doc.get(doc_attr, False) for job in jobs)
+
+    return true_for_all_docs
+
+
+class Project(flow.FlowProject):
+    """Create a workflow for simulating 2D Gaussian random walks."""
+
+    pass
+
+
+@Project.label
+def simulated(job):
+    return "positions" in job.data
+
+
+def all_simulated(*jobs):
+    return all(simulated(job) for job in jobs)
+
+
+# Create aggregator that combines all replicas with a single standard deviation
+std_aggregator = flow.aggregator.groupby("std")
+
+
+# Define all groups
+# -----------------
+plot = Project.make_group("aggregate-plot", aggregator=std_aggregator)
+analyze_and_plot = Project.make_group("post-processing", aggregator=std_aggregator)
+
+
+@Project.post(simulated)
+@Project.operation
+def simulate(job):
+    """Simulate a 2D random walk."""
+    # use job id as random seed
+    n_steps = job._project.doc.run_steps
+    generator = np.random.default_rng(int(job.id, 16))
+    # Generate all moves at once since they are independent
+    moves = generator.normal(job.sp.mean, job.sp.std, (n_steps, 2))
+    position = np.empty((n_steps + 1, 2), dtype=float)
+    position[0, :] = [0, 0]
+    # Store the results: cumsum aggregates all previous moves from the origin
+    position[1:] = np.cumsum(moves, axis=0)
+    job.data.positions = position
+
+
+@analyze_and_plot
+@std_aggregator
+@Project.pre(all_simulated)
+@Project.post(aggregate_cond_true("msd_analyzed"))
+@Project.operation
+def analyze_mean_squared_distance(*jobs):
+    """Compute and store the mean squared displacement for all std."""
+    msd = np.zeros((jobs[0]._project.doc.run_steps + 1), dtype=float)
+    for position in positions_generator(jobs):
+        msd += np.sum(np.multiply(position, position), axis=1)
+    msd /= len(jobs)
+    # Store msd in only first replica (job.sp.replica == 1)
+    get_first_replica(jobs).data["msd"] = msd
+    for job in jobs:
+        job.doc.msd_analyzed = True
+
+
+@Project.pre.true("msd_analyzed")
+@Project.pre(lambda job: job.sp.replica == 1)
+@Project.post.isfile("msd.png")
+@Project.operation
+def plot_mean_squared_distance(job):
+    """Plot the MSD for all std."""
+    with job.data:
+        msd = job.data.msd[:]
```

Comment:
You might need to index with an empty tuple to force it to copy the array into memory...?
```suggestion
        msd = job.data.msd[()]
```

Reply:
The `[:]` should be supported in modern `h5py` versions.

Reply:
In my understanding, `[:]` produces a view that expires when the file handle is closed, not a copy. The h5py 3.0 docs state: "Use `ds[()]` to read all data from any dataset." We've had a lot of difficulty with `[:]` producing views and breaking code in the past, so I'd like to be 100% sure about this.

https://docs.h5py.org/en/stable/whatsnew/3.0.html#breaking-changes-deprecations

Reply:
The following code on my local machine, in a Python 3.8.5 interpreter returns the following

```python
>>> with job.data:
...     print(type(job.data.positions[:]))
...
<class 'numpy.ndarray'>
>>>
```
My h5py version is 3.1.0

Reply:
I think this is the relevant example:
```
>>> import numpy as np
>>> job = pr.open_job({'a': 1})
>>> with job.data as d:
...     d['a'] = np.random.rand(3)
...
>>> job.data
H5Store(filename='workspace/42b7b4f2921788ea14dac5566e6f06d0/signac_data.h5')
>>> job.data['a']
>>> with job.data as d:
...     tmp = d['a'][:]
...
>>> tmp
array([0.50358323, 0.36469007, 0.57126965])  # This shows the data outside the context manager.
>>> with job.data as d:
...     tmp2 =  d['a']
...     print(tmp2)
...
<HDF5 dataset "a": shape (3,), type "<f8">
>>> tmp2
<Closed HDF5 dataset>  # Not loaded into memory.
```

Having said that it would be good to validate that this works in nested cases etc, although from my recollection it does.

Reply:
This might have been the reason I originally chose to use `[()]` over `[:]`: scalar values don't work with `[:]` but both arrays and scalars work with `[()]`. https://docs.h5py.org/en/stable/high/dataset.html#reading-writing-data
```python
import signac

project = signac.get_project()

project.data["scalar_key"] = 5

with project.data:
    scalar_view = project.data["scalar_key"][()]

print(scalar_view)  # prints 5

# Raises IndexError: invalid index to scalar variable.
with project.data:
    scalar_view = project.data["scalar_key"][:]
```

Reply:
Correct, scalars will not work with the colon syntax. For the purposes of this example (the MSD) I think a colon will be fine though. `[()]` and `[:]` should AFAIR be identical for array-like data.